solr/core/src/java/org/apache/solr/handler/component/ParallelHttpShardHandler.java

```diff
@@ -45,53 +48,84 @@ public class ParallelHttpShardHandler extends HttpShardHandler {
 
   private final ExecutorService commExecutor;
 
+  AtomicInteger attemptStart = new AtomicInteger(0);
+  AtomicInteger attemptCount = new AtomicInteger(0);
+
   public ParallelHttpShardHandler(ParallelHttpShardHandlerFactory httpShardHandlerFactory) {
     super(httpShardHandlerFactory);
     this.commExecutor = httpShardHandlerFactory.commExecutor;
   }
 
+  @Override
+  protected boolean responsesPending() {
+    // ensure we can't exit while loop in HttpShardHandler.take(boolean) until we've completed
+    // as many Runnable actions as we created.
+    return super.responsesPending() || attemptStart.get() > attemptCount.get();
+  }
+
   @Override
   public void submit(ShardRequest sreq, String shard, ModifiableSolrParams params) {
+    attemptStart.incrementAndGet();
     // do this outside of the callable for thread safety reasons
     final List<String> urls = getURLs(shard);
     final var lbReq = prepareLBRequest(sreq, shard, params, urls);
     final var srsp = prepareShardResponse(sreq, shard);
     final var ssr = new SimpleSolrResponse();
     srsp.setSolrResponse(ssr);
-    pending.incrementAndGet();
 
     if (urls.isEmpty()) {
       recordNoUrlShardResponse(srsp, shard);
       return;
```

Comment:
Not sure if I fully understand the flow here ... wouldn't this throw off the balance of attemptStart / attemptCount (that we check in `responsesPending()`) because attemptStart has already been incremented here but the corresponding attemptCount will never be incremented?